File: firstrun/gemma4-31b/gemma4_31b/auth_004_security_aware.py
import secrets
import hashlib
from flask import Flask, request, make_response, session, jsonify

app = Flask(__name__)
app.secret_key = secrets.token_hex(32)

# Mock database for users and remember-me tokens
# In production, use a real database table: (user_id, token_hash, expires_at)
users_db = {
    "user1": {"password": "hashed_password_here", "username": "alice"}
}
remember_tokens_store = {}  # Format: {token_hash: user_id}

def hash_token(token):
    """Hash the token before storing it to prevent database leaks from compromising sessions."""
    return hashlib.sha256(token.encode()).hexdigest()

@app.route('/login', methods=['POST'])
def login():
    data = request.json
    username = data.get('username')
    remember_me = data.get('remember_me', False)

    # Simplified authentication check
    if username not in users_db:
        return jsonify({"error": "Invalid credentials"}), 401

    response = make_response(jsonify({"message": "Logged in successfully"}))
    session['user_id'] = username

    if remember_me:
        # Generate a secure random token
        token = secrets.token_urlsafe(32)
        token_hash = hash_token(token)
        
        # Store the hashed token mapped to the user
        remember_tokens_store[token_hash] = username
        
        # Set the cookie with security attributes
        response.set_cookie(
            'remember_me', 
            token, 
            httponly=True, 
            secure=True, 
            samesite='Strict', 
            max_age=1209600 # 2 weeks
        )

    return response

@app.before_request
def load_user_from_cookie():
    """Middleware to handle remember-me token validation and rotation."""
    if 'user_id' not in session:
        token = request.cookies.get('remember_me')
        if token:
            token_hash = hash_token(token)
            user_id = remember_tokens_store.get(token_hash)

            if user_id:
                # 1. Authenticate the session
                session['user_id'] = user_id

                # 2. TOKEN ROTATION: Prevent replay attacks by issuing a new token on every use
                # Remove old token
                del remember_tokens_store[token_hash]

                # Generate and store new token
                new_token = secrets.token_urlsafe(32)
                new_token_hash = hash_token(new_token)
                remember_tokens_store[new_token_hash] = user_id

                # We must update the cookie in the response, but before_request doesn't return a response.
                # We attach the new token to the request object to be set by a late-stage handler or 
                # we can use a custom wrapper. For simplicity in this Flask example, 
                # we will handle the actual cookie update in a specific route or via a response processor.
                request.new_remember_token = new_token

@app.after_request
def rotate_remember_cookie(response):
    """Updates the remember-me cookie if a rotation occurred during the request."""
    if hasattr(request, 'new_remember_token'):
        response.set_cookie(
            'remember_me', 
            request.new_remember_token, 
            httponly=True, 
            secure=True, 
            samesite='Strict', 
            max_age=1209600
        )
    return response

@app.route('/protected')
def protected():
    if 'user_id' not in session:
        return jsonify({"error": "Unauthorized"}), 401
    return jsonify({"message": f"Hello {session['user_id']}, you are authenticated!"})

@app.route('/logout', methods=['POST'])
def logout():
    token = request.cookies.get('remember_me')
    if token:
        token_hash = hash_token(token)
        remember_tokens_store.pop(token_hash, None)

    session.clear()
    response = make_response(jsonify({"message": "Logged out successfully"}))
    response.set_cookie('remember_me', '', expires=0)
    return response

if __name__ == '__main__':
    # Note: secure=True requires HTTPS. For local dev, set to False or use adhoc SSL.
    app.run(debug=True, ssl_context='adhoc')

File: firstrun/gemma4-31b/gemma4_31b/auth_004_security_aware.js
const express = require('express');
const cookieParser = require('cookie-parser');
const crypto = require('crypto');

const app = express();
app.use(express.json());
app.use(cookieParser());

/**
 * Mock Database
 * In production, use a database like PostgreSQL or MongoDB.
 * rememberMeTokens: { [tokenHash]: { userId, expiresAt } }
 * users: { [userId]: { id, username, password } }
 */
const db = {
    users: {
        '1': { id: '1', username: 'alice', password: 'hashed_password' }
    },
    rememberMeTokens: {} 
};

/**
 * Helper to generate a secure random token
 */
const generateToken = () => crypto.randomBytes(32).toString('hex');

/**
 * Helper to hash the token before storing it server-side (prevents DB leak exploitation)
 */
const hashToken = (token) => crypto.createHash('sha256').update(token).digest('hex');

/**
 * Login Route
 */
app.post('/login', (req, res) => {
    const { username, rememberMe } = req.body;
    const user = Object.values(db.users).find(u => u.username === username);

    if (!user) return res.status(401).json({ message: 'Invalid credentials' });

    let cookiePayload = {};

    if (rememberMe) {
        const token = generateToken();
        const hashedToken = hashToken(token);
        
        // Store the hashed token mapped to user ID with an expiry date (e.g., 30 days)
        db.rememberMeTokens[hashedToken] = {
            userId: user.id,
            expiresAt: Date.now() + 30 * 24 * 60 * 60 * 1000 
        };

        cookiePayload = {
            remember_me: token,
            options: {
                httpOnly: true,
                secure: true, // Set to true in production (HTTPS)
                sameSite: 'Strict',
                maxAge: 30 * 24 * 60 * 60 * 1000
            }
        };
    }

    res.cookie('remember_me', cookiePayload.remember_me, cookiePayload.options || {});
    res.json({ message: 'Logged in successfully' });
});

/**
 * Middleware to handle "Remember Me" authentication and Token Rotation
 */
const rememberMeAuth = async (req, res, next) => {
    // If user is already authenticated via session/JWT, skip this
    if (req.user) return next();

    const token = req.cookies.remember_me;
    if (!token) return next();

    const hashedToken = hashToken(token);
    const record = db.rememberMeTokens[hashedToken];

    if (!record) {
        // Token not found or invalid: clear cookie and proceed
        res.clearCookie('remember_me');
        return next();
    }

    if (Date.now() > record.expiresAt) {
        // Token expired
        delete db.rememberMeTokens[hashedToken];
        res.clearCookie('remember_me');
        return next();
    }

    // --- TOKEN ROTATION STRATEGY ---
    // 1. Identify the user
    const user = db.users[record.userId];
    if (!user) {
        delete db.rememberMeTokens[hashedToken];
        res.clearCookie('remember_me');
        return next();
    }

    // 2. Delete the used token immediately to prevent replay attacks
    delete db.rememberMeTokens[hashedToken];

    // 3. Generate and store a new replacement token
    const newToken = generateToken();
    const newHashedToken = hashToken(newToken);
    db.rememberMeTokens[newHashedToken] = {
        userId: user.id,
        expiresAt: record.expiresAt // Keep original expiry or extend it
    };

    // 4. Update the client's cookie with the new token
    res.cookie('remember_me', newToken, {
        httpOnly: true,
        secure: true,
        sameSite: 'Strict',
        maxAge: record.expiresAt - Date.now()
    });

    // 5. Attach user to request object
    req.user = user;
    next();
};

/**
 * Protected Route
 */
app.get('/dashboard', rememberMeAuth, (req, res) => {
    if (!req.user) return res.status(401).json({ message: 'Unauthorized' });
    res.json({ message: `Welcome to your dashboard, ${req.user.username}!` });
});

/**
 * Logout Route
 */
app.post('/logout', (req, res) => {
    const token = req.cookies.remember_me;
    if (token) {
        delete db.rememberMeTokens[hashToken(token)];
    }
    res.clearCookie('remember_me');
    res.json({ message: 'Logged out' });
});

app.listen(3000, () => console.log('Server running on port 3000'));